io/src/main/java/com/github/jorgecastillo/kotlinandroid/io/runtime/ui/SuperHeroDetailActivity.kt

```diff
@@ -17,50 +18,51 @@ import kotlinx.android.synthetic.main.activity_detail.*
 
 class SuperHeroDetailActivity : AppCompatActivity(), SuperHeroDetailView {
 
-  companion object {
-    const val EXTRA_HERO_ID = "EXTRA_HERO_ID"
+    companion object {
+        const val EXTRA_HERO_ID = "EXTRA_HERO_ID"
 
-    fun launch(source: Context, heroId: String) {
-      val intent = Intent(source, SuperHeroDetailActivity::class.java)
-      intent.putExtra(
-          EXTRA_HERO_ID, heroId)
-      source.startActivity(intent)
+        fun launch(source: Context, heroId: String) {
+            val intent = Intent(source, SuperHeroDetailActivity::class.java)
+            intent.putExtra(
+                    EXTRA_HERO_ID, heroId)
+            source.startActivity(intent)
+        }
     }
-  }
 
-  override fun onCreate(savedInstanceState: Bundle?) {
-    super.onCreate(savedInstanceState)
-    setContentView(R.layout.activity_detail)
-  }
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_detail)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        intent.extras?.let {
+            val heroId = it.getString(
+                    EXTRA_HERO_ID)
+            unsafe { runNonBlocking({ Presentation.drawSuperHeroDetails(heroId, this@SuperHeroDetailActivity) }, {}) }
 
-  override fun onResume() {
-    super.onResume()
-    intent.extras?.let {
-      val heroId = it.getString(
-          EXTRA_HERO_ID)
-      Presentation.drawSuperHeroDetails(heroId, this).fix().unsafeRunAsync { }
-    } ?: closeWithError()
-  }
+        } ?: closeWithError()
+    }
 
-  private fun closeWithError() {
-    Toast.makeText(this, string.hero_id_needed, Toast.LENGTH_SHORT).show()
-  }
+    private fun closeWithError() {
+        Toast.makeText(this, string.hero_id_needed, Toast.LENGTH_SHORT).show()
+    }
 
-  override fun drawHero(hero: SuperHeroViewModel) = runOnUiThread {
-    collapsingToolbar.title = hero.name
-    description.text = hero.description.let { if (it.isNotEmpty()) it else getString(string.empty_description) }
-    headerImage.loadImageAsync(hero.photoUrl)
-  }
+    override fun drawHero(hero: SuperHeroViewModel) = runOnUiThread {
```

Comment:
This should be a suspend function now, and jump to the main thread from an `fx` before calling it.

Calling UI on `subscribe` or `unsafeRunXXX` isn't a valid approach anymore, we're wrapping _all_ side-effects :D

Reply:
Yep this is old leftovers, not used anymore. Need to remove it. It's getting off the Main thread and back all together in the services layer (DataSource.kt). I'm fine running everything else in the architecture on Main thread since it's not computationally expensive, so I'm just leaving Main and back on those given spots.